### PR TITLE
dts: nordic: Use NRF_DEFAULT_IRQ_PRIORITY instead of hard-coded values

### PR DESCRIPTION
**Pull request by anangl**
This is an alternative approach to solving the problem addressed also by #19703 and described in https://github.com/zephyrproject-rtos/zephyr/pull/19703#issuecomment-539929845.

---

Instead of hard-coded values for interrupt priorities in DT nodes, use
a macro that additionally can be overridden at the board or application
level. This allows, for instance, changing the interrupt priorities for
all but one peripheral without using a lengthy overlay file.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

### CODE REVIEW

File: dts/arm/nordic/nrf51822.dtsi
Comment (cvinayak) [inline dts/arm/nordic/nrf51822.dtsi]: Use the lowest priority as default, which is 3 IIRC for nRF51

File: dts/arm/nordic/nrf52810.dtsi
Comment (cvinayak) [inline dts/arm/nordic/nrf52810.dtsi]: Use the lowest priority as default, which is 6 IIRC for nRF52, as highest priority is used for Zero Latency IRQs.

Similarly for rest of the nRF52, nRF53 and nRF91 series

**Review by cvinayak — DISMISSED** [review]

File: dts/arm/nordic/nrf51822.dtsi
Comment (anangl) [reply dts/arm/nordic/nrf51822.dtsi]: I am not sure we should do this globally, as this can potentially affect existing applications. Currently this PR only makes changing of default priority easier, the existing values are not modified, except for GPIOTE, which for unknown reason had priority 5 set on nRF52, and that got propagated on nRF91 and nRF53 (on nRF51 it was 1, not 3).
Perhaps the priorities should be lowered only where it is actually needed, e.g. in BLE-related samples, as suggested by @ioannisg in https://github.com/zephyrproject-rtos/zephyr/pull/19703#issuecomment-542133475.

What are the others' opinions regarding this?

File: dts/arm/nordic/nrf51822.dtsi
Comment (ulfalizer) [inline dts/arm/nordic/nrf51822.dtsi]: Not hardcore about it, but maybe the `DT_` prefix could be left out, so that `DT_` is reserved for macros automatically generated by the devicetree scripts, similar to how `CONFIG_` is reserved for Kconfig.

Comment (ulfalizer) [reply dts/arm/nordic/nrf51822.dtsi]: `DEFAULT_DT_IRQ_PRIORITY` might work too.

Comment (anangl) [reply dts/arm/nordic/nrf51822.dtsi]: I initially thought about not prefixing this macro with `DT_` but then I saw https://github.com/zephyrproject-rtos/zephyr/blob/d4ba8fff66bd39f413a853956aafd4609041cec3/dts/common/mem.h#L10
and got a bit confused. Is it stated somewhere that `DT_` prefix is only for the generated stuff (and related fixups)?

I like the `DEFAULT_DT_IRQ_PRIORITY` proposal. Can use it if there are no objections.

**Review by SebastianBoe — APPROVED** [review]

**Review by ioannisg — COMMENTED** [review]
Can you do a minor cleanup, @anangl ?
Put the #ifdef check inside nrf5_common.dtsi, so it is not copied in all different .dtsi headers.
You might need to do some minor re-work (e.g. include nrf5_common.dtsi in nrf5340 and nrf91, and perhaps rename to nrfx_common.dtsi) to manage to do the above.

File: dts/arm/nordic/nrf51822.dtsi
Comment (ulfalizer) [reply dts/arm/nordic/nrf51822.dtsi]: > Is it stated somewhere that `DT_` prefix is only for the generated stuff (and related fixups)?

Nothing written down. Had forgotten that fixup files also #define `DT_` stuff. Maybe that could be cleaned up a bit at some point too.

Guess there might be less risk of confusion for function-like macros like `DT_SIZE_K`, though I wouldn't mind changing that as well.

**Review by pabigot — CHANGES_REQUESTED** [review]
Please rename the include file to something like `arm/nordic_nrf/override.dtsi` to make it clear that the content of the file is intended for inclusion into devicetree sources.  This mediates some concerns raised in #20910 and legitimizes the use of `DT_DEFAULT_IRQ_PRIORITY` because that name will never appear in the same C language context as the generated `DT_foo` symbols with which it might conflict.  , Also I would expect other vendors would want to do something like this, so the path should be consistent with the way vendor HALs are currently identified.

With that step taken to avoid namespace conflict I have no issues with using `DT_DEFAULT_IRQ_PRIORITY` in devicetree source this way.

My preference is to not change priorities from their current default, certainly not in this PR.  The defaults could/should be reviewed and changed where necessary, e.g. by introducing specific macros like `DT_DEFAULT_NORDIC_GPIOTE_IRQ_PRIORITY` to the set of overridable values.

I also agree with moving all the include/ifdef boilerplate into an existing common dtsi file, if that can be done.

**Comment by ulfalizer** [comment]
> This mediates some concerns raised in #20910 and legitimizes the use of `DT_DEFAULT_IRQ_PRIORITY` because that name will never appear in the same C language context as the generated `DT_foo` symbols with which it might conflict.

Was more about being confusing to see `DT_` on something that's not generated by the scripts for me. Unlikely that it'd clash.

File: dts/arm/nordic/nrf51822.dtsi
Comment (cvinayak) [reply dts/arm/nordic/nrf51822.dtsi]: > potentially affect existing applications

This is not an acceptable reasoning for me to select a default priority that is not highest or lowest in an SoC.

- Defaults can be to have highest priority and the use of Bluetooth controller would then change the defaults of all IRQs except the ones used by Bluetooth to a lower value. This is even more confusing to user of other peripherals.
- can be to have lowest priority and only the Bluetooth controller used IRQs are set to the required priority

**Review by pabigot — APPROVED** [review]

**Comment by anangl** [comment]
As requested:
- moved the common include/ifdef stuff to `nrf_common.dtsi` that replaced `nrf5_common.dtsi`
- changed override file's extension to `.dtsi`
- used `DT_DEFAULT_IRQ_PRIORITY` (the name is still to be discussed/agreed on) only instead of priorities that were already set to `1` (this means almost all nodes, `gpiote` in all but nRF51 and `rng` in nRF5340 are the exceptions), so that this PR would not modify the existing priorities at all

**Comment by cvinayak** [comment]
@anangl can we resume discussions? (regarding use lowest IRQ priority levels as default)
@pabigot do you have any objections to use lowest interrupt priority level as default?

**Comment by anangl** [comment]
> @anangl can we resume discussions? (regarding use lowest IRQ priority levels as default)

I agree with @pabigot's comment in https://github.com/zephyrproject-rtos/zephyr/pull/20894#pullrequestreview-320845105:
> My preference is to not change priorities from their current default, certainly not in this PR. The defaults could/should be reviewed and changed where necessary, e.g. by introducing specific macros like `DT_DEFAULT_NORDIC_GPIOTE_IRQ_PRIORITY` to the set of overridable values.

that the default priorities should not be changed in this PR, as this one is only about making them configurable. Let's do it in a follow-up PR.

**Comment by anangl** [comment]
Rebased on master.

**Comment by anangl** [comment]
Applied the same modifications in the recently added files (`nrf52805.dtsi ` and `nrf52820.dtsi`).

**Review by carlescufi — COMMENTED** [review]
I personally think this is a good approach, but I will leave the others that deal with DT to do the approving. Especially since we should think whether this is an approach that can be generalized and reused by all other SoC vendors.

**Comment by cvinayak** [comment]
Please create two follow up GH issue related to GPIOTE should also use the new DT define, and the other to follow up to change  default interrupt priority value.

**Comment by cvinayak** [comment]
Please create two follow up GH issue related to GPIOTE should also use the new DT define, and the other to follow up to change  default interrupt priority value.

**Comment by pabigot** [comment]
git grep interrupts '*/nrf*.dts*' | grep -v DEFAULT

shows about 20 interrupts properties that weren't switched.  Some of those are 5, but some are 1.  Should they be converted to the default also?

**Comment by anangl** [comment]
> ```
> git grep interrupts '*/nrf*.dts*' | grep -v DEFAULT
> ```
> 
> shows about 20 interrupts properties that weren't switched. Some of those are 5, but some are 1. Should they be converted to the default also?

Yes. I forgot to look for nodes that have been added in files that were already modified by this PR. My bad. Thanks for catching this @pabigot!
Now only GPIOTE nodes are left unmodified.

**Review by mbolivar-nordic — COMMENTED** [review]
I think this ought to be named something like NRF_DEFAULT_IRQ_PRIORITY (or feel free to pick another name that doesn't imply it's a system-wide default from devicetree).

I think if we leave it as-is, people will thing it is supported by other SoCs, but that doesn't seem to be the case.

@galak?

**Comment by pabigot** [comment]
> I think this ought to be named something like NRF_DEFAULT_IRQ_PRIORITY (or feel free to pick another name that doesn't imply it's a system-wide default from devicetree).
> 
> I think if we leave it as-is, people will thing it is supported by other SoCs, but that doesn't seem to be the case.

Maybe `SOC_DEFAULT_IRQ_PRIORITY`? If part of the goal is to move to using this idiom across all target platforms I don't think it should be Nordic-specific by name; that would also complicate use in overlays (assuming such use is meaningful).  I doubt we'd get all targets to use the same default (highest/lowest), so maybe `DEFAULT_IRQ_PRIORITY` isn't ideal.

File: dts/arm/nordic/nrf_common.dtsi
Comment (galak) [inline dts/arm/nordic/nrf_common.dtsi:16]: This should be something like DT_NRF_DEFAULT_IRQ_PRIORITY, since this is not a generic feature.

**Review by galak — CHANGES_REQUESTED** [review]
DT_DEFAULT_IRQ_PRIORITY should be renamed to something w/NRF in it.

**Comment by galak** [comment]
> I think this ought to be named something like NRF_DEFAULT_IRQ_PRIORITY (or feel free to pick another name that doesn't imply it's a system-wide default from devicetree).
> 
> I think if we leave it as-is, people will thing it is supported by other SoCs, but that doesn't seem to be the case.
> 
> @galak?

agreed, should have some NRF qualifier.

**Comment by anangl** [comment]
Renamed `DT_DEFAULT_IRQ_PRIORITY` to `NRF_DEFAULT_IRQ_PRIORITY`.

**Comment by anangl** [comment]
@mbolivar-nordic @galak ping

File: dts/arm/nordic/nrf_common.dtsi
Comment (ioannisg) [inline dts/arm/nordic/nrf_common.dtsi:16]: Is seems a bit weird that we start defining preprocessor macros in .dtsi headers.
Is this going to be an accepted policy?

@galak @mbolivar-nordic ^^

If yes, I am fine with this, too.

Comment (pabigot) [reply dts/arm/nordic/nrf_common.dtsi:16]: It has precedent in Linux as well as Zephyr.  I think a case can be made that they should be `.h` files (even though I believe I formerly requested it be `.dtsi`), but IMO that should not be a blocker for this until it's changed uniformly in Zephyr.

Comment (ioannisg) [reply dts/arm/nordic/nrf_common.dtsi:16]: Whether it is .h or .dtsi is not a blocker, of course.
But I want a wider consensus that we will be accepting the use of pre-processor macros/directives in DTS.

**Review by pabigot — APPROVED** [review]
sorry, thought I'd approved this.

File: dts/arm/nordic/nrf_common.dtsi
Comment (pabigot) [reply dts/arm/nordic/nrf_common.dtsi:16]: We've accepted the use of preprocessor directives in #27719, so in my mind it's only relevant for defines.  https://zephyrproject.slack.com/archives/C18PLHN8H/p1598280951131500 has a discussion that relates to another discussion; the conclusion there is that trying to make a `.h` file work for both code and devicetree is problematic.

But if you want a wider consensus I suggest opening an issue.  Nobody's going to find these review comments when trying to research how a decision was reached.

Comment (ioannisg) [reply dts/arm/nordic/nrf_common.dtsi:16]: Nope, I copied the DT maintainers, above, with which I seek consensus, no need to file an issue.

**Review by ioannisg — APPROVED** [review]

File: dts/arm/nordic/nrf_common.dtsi
Comment (mbolivar-nordic) [reply dts/arm/nordic/nrf_common.dtsi:16]: I'm fine with it if @galak is.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
@galak please revisit

**Comment by anangl** [comment]
Rebased on master, to resolve the conflict in `nrf5_common.dtsi`.

**Comment by anangl** [comment]
Updated the PR, to accommodate the recent changes in master (especially in `nrf5_common.dtsi`).

**Comment by anangl** [comment]
@galak Would you mind taking another look? I addressed your comment.

**Comment by ioannisg** [comment]
@galak, ping for finalizing the review of this one !

File: dts/arm/nordic/nrf_common.dtsi
Comment (ioannisg) [inline dts/arm/nordic/nrf_common.dtsi:19]: Since there's not a central place to disable the SysTick node (i.e. for all nRF SoCs), it would be nice to add a minor explanatory comment above this line, to justify why we disable the SyStick node. I.e. "Nordic SoCs rely by default on the RTC for system clock implementation, so the SysTick node is not to be enabled"

**Comment by anangl** [comment]
Rebased, to resolve merge conflicts.

**Comment by anangl** [comment]
Updated the `qspi` node that has been recently added in `nrf5340_cpuapp_peripherals.dtsi`.

**Comment by anangl** [comment]
Rebased once more, to resolve merge conflicts.

File: dts/arm/nordic/nrf_common.dtsi
Comment (anangl) [reply dts/arm/nordic/nrf_common.dtsi:19]: Added.

**Comment by anangl** [comment]
Rebased again. Added the explanatory comment suggested by @ioannisg in https://github.com/zephyrproject-rtos/zephyr/pull/20894#pullrequestreview-510225911.

**Comment by anangl** [comment]
One more rebase, to make compliance checks pass.

**Review by cvinayak — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**
